tools/topology/sof-cml-rt1011-rt5682.m4

```diff
@@ -62,11 +62,18 @@ PCM_PLAYBACK_ADD(Speakers, 5, PIPELINE_PCM_7)
 #SSP 1 (ID: 6)
 #Use BCLK delay in SSP_CONFIG_DATA only on supporting version
 
-DAI_CONFIG(SSP, SSP1_INDEX, 6, SSP1_NAME,
+ifdef(`CML_RT1015', 
+`DAI_CONFIG(SSP, 1, 6, SSP1-Codec,
+        SSP_CONFIG(I2S, SSP_CLOCK(mclk, 24000000, codec_mclk_in),
+                SSP_CLOCK(bclk, 3072000, codec_slave),
+                SSP_CLOCK(fsync, 48000, codec_slave),
+                SSP_TDM(2, 32, 3, 3),
+                SSP_CONFIG_DATA(SSP, 1, 24)))',
```

Comment:
does this work? How can you generate 32 bits from a 24 MHz clock? That is not a multiple of 48 kHz?

Reply:
They are from: https://github.com/thesofproject/sof/blob/master/tools/topology/platform/intel/jsl-rt1015.m4
and do work from our test.

Reply:
Bit clock comes from SSP M/N counter so it’s ok to use MCLK 24MHz. 
Bclk/fsync=3072000/48000=64 bits per fsync
TDM setting: 2 slots 32 bits => 64 bits per fsync
The setting is match.

Reply:
The M/N counters don't guarantee a fixed duty cycle and introduce clock jitter by design.

The recommendation is NOT to use them, unless it's proven that the codec/amp tolerates variations. if you can it's much better to use integer divisions and 25 bits/slots.

Reply:
I've got response from realtek, they prefer I2S 64fs for this amp on all platforms. They don't have concern for the clock jitter. I think it's based on the result on JSL which is also running a 24MHz mclk.

Reply:
@brentlu IIRC JSL is derived from ICL, not CML,  and it has the 24.576MHz PLL so your results are likely not quite right. You cannot compare results unless the clock tree is identical.

Reply:
Sorry for the mistake. I checked the clk table, it seems CML only supports 24MHz XTAL and 96MHz PLL as source. Since this amp has 64fs limitation, we cannot bypass the m/n counter.

Reply:
I left the messages which sources from our HW engineer of Realtek. 
The 100fs and 128fs already verified on CML+ALC1011, there should be no concern on Realtek AMP for CML 128/64fs jitter.

Reply:
@shumingfan the RT1011 configuration uses the 25 bits per slot configuration with 4 slots, so 100xfs. Is there a reason why we cannot use the same with RT1015? Is the IP different? If not I suggest we use the same topology...

Reply:
@plbossart ALC1015 will also support auto mode which fixed on I2S 64fs 48k 24bit BCLK=3.072M in the feature.
So we would like to fix the I2S format on all platforms to keep it simple and reduce changing effort for topology.
We also upstream 2 different code rt1015.c and rt1015p.c for I2C mode and auto mode application.
If there’s no obvious concern please let us using  “I2S 64fs 48k 24bit BCLK=3.072M” on all platforms.

Reply:
@shumingfan you missed the point that the 64fs does add jitter on CML. If you can use the 25 bit slot across all amplifiers, it's a much better solution. The 64 fs and M/N dividers should be the last thing to look into, not the default.

Reply:
@plbossart OK, I added the TDM slot configuration in RT1015 driver and sent the patch to ODM. I think we could use the same topology as rt1011+rt5682i. We are working on it.

Reply:
Sounds good @shumingfan, I really appreciate your work!